pallene/cfg.lua

```diff
@@ -0,0 +1,276 @@
+-- Copyright (c) 2020, The Pallene Developers
+-- Pallene is licensed under the MIT license.
+-- Please refer to the LICENSE and AUTHORS files for details
+-- SPDX-License-Identifier: MIT
+
+local Set = require "pallene.set"
+
+local CFG = {}
+--[[
+    A control flow graph is a representation of all paths 
+    that might be traversed through a program execution. 
+    
+    The nodes represent commands that use or alter data 
+    and the edges represent control instructions.
+
+    In this module, each node is defined by two sets of node indices (from and to)
+    and a list of commands. These commands are pointers to nodes in the ir module.
+
+    All the commands of the list are one of the listed in the ir module
+    except for the control flow ones (Nop, Seq, Return, Break, Loop, If and For).
+
+    Values aren't nodes.
+
+--]]
+
+
+--[[
+    Graph building uses 3 local functions: 
+    new_node, new_edge and build.
+    build is called on CFG.new.
+--]]
+
+-- creates a new node, should only be called to non-control commands.
+local function make_node(self, cmd,current_node)
+    table.insert(self,{
+        cmd = cmd, --TODO a list of commands
+        to={},
+        from={}
+    })
+    local new_node_id = #self
+    local new_node = self[new_node_id]
+    new_node.from = Set.copy(current_node)
+    Set.iforeach(new_node.from,function(_,from)
+        Set.insert(self[from].to,new_node_id)
+    end)
+    return {new_node_id}
+end
+-- creates a new edge from 'from' to 'to'
+local function make_edge(self,to,from)
+    if to and from then
+        Set.insert(self[from].to,to)
+        Set.insert(self[to].from,from)
+    end
+end
+
+-- builds the graph and calls new_node and new_edge.
+-- it receives a node to analyze, the current node list and the current loop.
+-- it returns an updated current node list and the current loop.
+
+-- usually, the current_node is a set with a single element, an index, 
+-- representing the last node visited or the size of the cfg.
+-- but when there is a branch meeting, such after ifs or in before loops,
+-- it becomes necessary to join the branches, having a list of current nodes
+
+-- current_loop is modelled as a stack, having a set of indexes as the element.
```

Comment:
It isn't modelled, it is a stack :) That said, it is a index to what exactly?

But do we really need a stack? In uninitialized.lua we use the recursion stack instead of an explicit stack. The function parameter only refers only to the innermost loop that we are a part of.